pkg/controllermanager/apis/config/types.go

```diff
@@ -77,6 +77,44 @@ type CloudProfileControllerConfiguration struct {
 	// ConcurrentSyncs is the number of workers used for the controller to work on
 	// events.
 	ConcurrentSyncs int
+	// KubernetesVersionManagement configures the version policy that applies to
+	// Kubernetes versions in the CloudProfile
+	KubernetesVersionManagement *KubernetesVersionManagement
+	// MachineImageVersionManagement configures the version policy that applies to
+	// MachineImage versions in the CloudProfile
+	MachineImageVersionManagement *MachineImageVersionManagement
+}
+
+// KubernetesVersionManagement configures the version policy that applies to
+// Kubernetes versions in the CloudProfile
+// You can read more about it here: https://github.com/gardener/gardener/blob/master/docs/proposals/05-versioning-policy.md
+type KubernetesVersionManagement struct {
+	// Enabled defines whether the KubernetesVersionManagement is enabled
+	Enabled bool
+	// SupportedKubernetesVersions is the amount of minor Kubernetes versions that are considered to be "supported"
+	// refers to versions existing in the CloudProfile
+	// defaults to 3 as this is common practice in the Kubernetes Community
+	// e.g Versions in CloudProfile: 1.17, 1.15, 1.14, 1.13 -> Supported: 1.17 & 1.15 & 1.14, Unsupported: 1.13
+	SupportedKubernetesVersions *int
+	// DeprecationDurationSupportedVersion is the time until a deprecated Kubernetes patch version
+	// of a supported minor version expires
+	// defaults to 4 months (with each 30 days)
+	DeprecationDurationSupportedVersion *metav1.Duration
+	// DeprecationDurationUnsupportedVersion is the time until a deprecated Kubernetes patch version
+	// of an unsupported minor version expires
+	// defaults to 1 month (with 30 days)
+	DeprecationDurationUnsupportedVersion *metav1.Duration
+}
+
+// MachineImageVersionManagement configures the version policy that applies to
+// MachineImage versions in the CloudProfile
+// You can read more about it here: https://github.com/gardener/gardener/blob/master/docs/proposals/05-versioning-policy.md
+type MachineImageVersionManagement struct {
```

Comment:
Because there are different settings involved e.g the different deprecation times for k8 versions with a supported or unsupported minor.

Reply:
Let's say a version is just a number and we want to be able to apply a policy to it.

At the moment, a policy is tightly coupled to the component the version refers to, i.e. policy features for K8s and policy features for machine image versions. The question is if we can have a more generic implementation of version mangement.

For instance, if you renamed `SupportedKubernetesVersions` --> `SimultaneouslyMaintaintedVersions` machine image versions could support the same policy features as K8s does.

Reply:
If I understand correctly, you suggest implementing "version policies" that can be applied to any kind of version instead of specialised handling for different kinds of versions (like machine image version).

You would you apply a "version policy" to a version kind (like kubernetes version)
through a generic config option like "SimultaneouslyMaintaintedVersions". 

I am having a hard time coming up with possible  "version policies" - do you think of something like
a)  expiration date for latest patch version of a minor version 
b) set expiration date at all
c)  SimultaneouslyMaintaintedVersions
-  ? 

I think the main upside to that approach would be greater flexibility for operators of Gardener to configure the version management. I am though not sure how to concretely implement that. Maybe you have spent some thoughts on it already.

However I would also like to ask:
- do we need such a generic approach, considering that we only have two relevant versions in a Shoot cluster for the foreseeable future?
- Do "version policies" a) and b) mentioned above do even make sense for machine image versions? So far it has been the assumption that versions can always be update to the latest version - a minor update is not critical.
- I think it is imperative that Kubernetes versions have a "version policy" that treats deprecation of latest patch versions of a minor version a certain way. Meaning we have a mandatory "version policy". Would you enforce that in an admission plugin?

Reply:
I wasn't about to bring this kind of complexity to Gardener. The proposal is rather to not make a difference between machine image and Kubernetes versions. At the moment the difference is `SupportedKubernetesVersions` which allows to (optionally) automatically sort out versions which aren't maintained any more. So why not also offering this for machine image versions? The feature is optional and can be used whenever adequate.

My thought is that having a unified version treatment makes API usage as well as code maintenance easier.